File: src/swingjs/JSAppletPanel.java
package swingjs;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.Enumeration;
import java.util.Hashtable;

import jsjava.applet.Applet;
import jsjava.applet.AppletContext;
import jsjava.applet.AppletStub;
import jsjava.awt.BorderLayout;
import jsjava.awt.Dimension;
import jsjava.awt.Font;
import jsjava.awt.Graphics;
import jsjava.awt.Image;
import jsjava.awt.Panel;
import jsjava.awt.Toolkit;
import jsjavax.swing.JApplet;
import jssun.applet.AppletEvent;
import jssun.applet.AppletEventMulticaster;
import jssun.applet.AppletListener;
import swingjs.api.HTML5Applet;
import swingjs.api.HTML5Canvas;
import swingjs.api.JSInterface;
import jsjavax.swing.JPanel;

/**
 * JSAppletPanel 
 * 
 * SwingJS class to start an applet. Note that this must be a JApplet,
 * not just java.awt.Applet. The implementation (for now) does not allow
 * "mixed" contents -- That is, no non-Swing Applet components are allowed.
 * 
 * The basic start up in JavaScript involves:
 * 
 * Clazz.loadClass("swingjs.JSAppletPanel"); 
 * var _appletPanel = new JSAppletPanel(viewerOptions);
 * _appletPanel.start();
 * 
 * where viewerOptions holds critical information needed to create this applet
 * 
 * 
 * @author Bob Hanson
 * 
 */
public class JSAppletPanel extends Panel implements AppletStub, AppletContext,
		JSInterface {

	private Hashtable params;
	/*
	 * the JavaScript SwingJS._Applet object
	 */
	public HTML5Applet html5Applet;

	public String fullName;
	public String appletCodeBase;
	public String appletIdiomaBase;
	public String appletDocumentBase;

	public int maximumSize = Integer.MAX_VALUE;
	public String appletName;
	public String syncId;
	public boolean testAsync;
	public boolean async;
	public String strJavaVersion;
	public Object strJavaVendor;
	public Object display;
	private HTML5Canvas canvas;
	private JSGraphics2D jsgraphics;
	private JApplet applet;

	// /// AppletPanel fields //////

	/**
	 * The initial applet size.
	 */
	Dimension defaultAppletSize = new Dimension(10, 10);

	/**
	 * The current applet size.
	 */
	Dimension currentAppletSize = new Dimension(10, 10);

	private int nextStatus;

	/* applet event ids */
	public final static int APPLET_UNINITIALIZED = 0; // SwingJS
	public final static int APPLET_LOAD = 1;
	public final static int APPLET_INIT = 2;
	public final static int APPLET_START = 3;
	public final static int APPLET_READY = 35; // SwingJS
	public final static int APPLET_STOP = 4;
	public final static int APPLET_DESTROY = 5;
	public final static int APPLET_QUIT = 6;
	public final static int APPLET_ERROR = 7;
	public final static int APPLET_DISPOSE = 75; // SwingJS

	/* send to the parent to force relayout */
	public final static int APPLET_RESIZE = 51234;

	/*
	 * sent to a (distant) parent to indicate that the applet is being loaded or
	 * as completed loading
	 */
	public final static int APPLET_LOADING = 51235;
	public final static int APPLET_LOADING_COMPLETED = 51236;

	/**
	 * The current status. One of: APPLET_DISPOSE, APPLET_LOAD, APPLET_INIT,
	 * APPLET_START, APPLET_STOP, APPLET_DESTROY, APPLET_ERROR.
	 */
	private int status = APPLET_UNINITIALIZED;

	private AppletListener listeners;

	/**
	 * SwingJS initialization is through a Hashtable provided by the page
	 * JavaScript
	 * 
	 * After the applet is instantiated is the opportunity to add a listener using
	 * setAppletListener(x), where x.appletStateChanged(AppletEvent evt) exists
	 * 
	 * next command on page should be appletPanel.start();
	 * 
	 * @param params
	 */
	public JSAppletPanel(Hashtable params) {
		setPanel();
		set(params);
	}

	private void set(Hashtable params) {
		System.out.println("JSAppletPanel initializing");
		this.params = params;
		htmlName = JSUtil.split("" + getParameter("name"), "_object")[0];
		appletName = JSUtil.split(htmlName + "_", "_")[0];
		// should be the same as htmlName; probably should point out that applet
		// names cannot have _ in them.

		syncId = getParameter("syncId");
		fullName = htmlName + "__" + syncId + "__";
		params.put("fullName", fullName);
		Object o = params.get("codePath");
		if (o == null)
			o = "../java/";
		appletCodeBase = o.toString();
		appletIdiomaBase = appletCodeBase.substring(0,
				appletCodeBase.lastIndexOf("/", appletCodeBase.length() - 2) + 1)
				+ "idioma";
		o = params.get("documentBase");
		appletDocumentBase = (o == null ? "" : o.toString());
		if (params.containsKey("maximumSize"))
			Math.max(((Integer) params.get("maximumSize")).intValue(), 100);
		async = (testAsync || params.containsKey("async"));
		HTML5Applet applet = null;
		String javaver = "?";
		/**
		 * @j2sNative
		 * 
		 *            if(self.Jmol) { applet =
		 *            Jmol._applets[this.htmlName.split("_object")[0]]; javaver =
		 *            Jmol._version; }
		 * 
		 * 
		 */
		{
		}
		html5Applet = applet;
		strJavaVersion = javaver;
		strJavaVendor = "Java2Script/Java 1.6 (HTML5)";
		// String platform = (String) params.get("platform");
		// if (platform != null && platform.length() > 0)
		// apiPlatform = (GenericPlatform) Interface.getInterface(platform);
		display = params.get("display");

		threadGroup = new JSThreadGroup(appletName);
		myThread = new JSAppletThread(this, threadGroup, appletName);
		((JSThreadGroup) threadGroup).setHtmlApplet(html5Applet);
		/**
		 * @j2sNative
		 * 
		 *            Jmol._applets[this.appletName + "_thread"] =
		 *            jsjava.lang.Thread.thisThread = this.myThread; 
		 * 
		 */
		{
		}
		appContext = JSToolkit.createNewAppContext();
		Toolkit.getDefaultToolkit();
		System.out.println("JSAppletPanel initialized");
	}

	public void start() {
		if (status == APPLET_UNINITIALIZED)
			myThread.start();
		else
			showStatus("already started");
	}

	synchronized public void addAppletListener(AppletListener l) {
		listeners = AppletEventMulticaster.add(listeners, l);
	}

	synchronized public void removeAppletListener(AppletListener l) {
		listeners = AppletEventMulticaster.remove(listeners, l);
	}

	/**
	 * Dispatch event to the listeners..
	 */
	public void dispatchAppletEvent(int id, Object argument) {
		if (listeners != null) {
			AppletEvent evt = new AppletEvent(this, id, argument);
			listeners.appletStateChanged(evt);
		}
	}

	private String getCode() {
		return getParameter("code");
	}

	private HTML5Canvas getCanvas() {
		if (canvas == null)
			canvas = html5Applet._getHtml5Canvas();
		return canvas;
	}

	// ///////// AppletStub ////////////////

	@Override
	public boolean isActive() {
		return true;
	}

	@Override
	/**
	 * Is called when the applet wants to be resized.
	 */
	public void appletResize(int width, int height) {
		final Dimension currentSize = new Dimension(currentAppletSize.width,
				currentAppletSize.height);
		currentAppletSize.width = width;
		currentAppletSize.height = height;
		applet.setBounds(0, 0, getWidth(), getHeight());
		applet.getRootPane().setBounds(0, 0, getWidth(), getHeight());
		applet.getContentPane().setBounds(0, 0, getWidth(), getHeight());
		((JPanel) applet.getContentPane()).revalidate();
		dispatchAppletEvent(APPLET_RESIZE, currentSize);
	}

	@Override
	public URL getDocumentBase() {
		try {
			return new URL((String) params.get("documentBase"));
		} catch (MalformedURLException e) {
			return null;
		}
	}

	@Override
	public URL getCodeBase() {
		try {
			return new URL((String) params.get("codeBase"));
		} catch (MalformedURLException e) {
			return null;
		}
	}

	@Override
	public String getParameter(String name) {
		String s = (String) params.get(name);
		System.out.println("get parameter: " + name + " = " + s);
		return s;
	}

	@Override
	public AppletContext getAppletContext() {
		return this;
	}

	// // Panel (Component) ////

	@Override
	public int getHeight() {
		return this.html5Applet._getHeight();
	}

	@Override
	public int getWidth() {
		return this.html5Applet._getWidth();
	}

	@Override
	public void setBounds(int x, int y, int width, int height) {
		reshape(x, y, width, height); // straight to component
		currentAppletSize.width = width;
		currentAppletSize.height = height;
	}

	// /// AppletContext /////

	@Override
	public Image getImage(URL url) {
		// TODO Auto-generated method stub
		return null;
	}

	@Override
	public Applet getApplet(String name) {
		Applet applet = null;
		/**
		 * @j2sNative
		 * 
		 *            applet = Jmol._applets[name]; applet && (applet =
		 *            applet._applet);
		 */
		{

		}
		return applet;
	}

	@Override
	public Enumeration<Applet> getApplets() {
		// not supported for now
		return null;
	}

	@Override
	public void showDocument(URL url) {
		/**
		 * @j2sNative window.open(url.toString());
		 */
		{
			System.out.println(url);
		}
	}

	@Override
	public void showDocument(URL url, String target) {
		/**
		 * @j2sNative window.open(url.toString(), target);
		 */
		{
			System.out.println(url);
		}
	}

	@Override
	public void showStatus(String status) {
		JSToolkit.log(status);
		/**
		 * @j2sNative
		 * 
		 *            Clazz._LoaderProgressMonitor.showStatus(status, true);
		 */
		{
			System.out.println(status);
		}
	}

	// ///////// javajs.api.JSInterface ///////////
	//
	// methods called by page JavaScript
	//
	//

	@Override
	public int cacheFileByName(String fileName, boolean isAdd) {
		// TODO Auto-generated method stub
		return 0;
	}

	@Override
	public void cachePut(String key, Object data) {
		// TODO Auto-generated method stub
	}

	@Override
	public void destroy() {
		// TODO Auto-generated method stub
	}

	@Override
	public String getFullName() {
		return fullName;
	}

	@Override
	public void openFileAsyncSpecial(String fileName, int flags) {
		// TODO Auto-generated method stub
	}

	@Override
	public boolean processMouseEvent(int id, int x, int y, int modifiers,
			long time) {
		// TODO Auto-generated method stub
		return false;
	}

	@Override
	public void processTwoPointGesture(float[][][] touches) {
		// TODO Auto-generated method stub
	}

	@Override
	public void setDisplay(HTML5Canvas canvas) {
		this.canvas = canvas;
	}

	@Override
	public void setScreenDimension(int width, int height) {
		setGraphics(jsgraphics = null);
		//resize(width, height);
		if (applet != null)
			applet.resize(width, height);
	}

	@Override
	public boolean setStatusDragDropped(int mode, int x, int y, String fileName) {
		// TODO Auto-generated method stub
		return false;
	}

	@Override
	public void startHoverWatcher(boolean enable) {
		// TODO Auto-generated method stub

	}

	/**
	 * @j2sOverride
	 */
	@Override
	public void paint(Graphics g) {
		// Note that this "Panel" is never painted.
		// This class simply maintains valuable information for applet loading.
		// Here we go straight to the contentPane and paint that.
		applet.getContentPane().paint(setGraphics(g));
	}

	/**
	 * SwingJS will deliver a null graphics here.
	 * 
	 * @param g
	 * @return
	 */
	private Graphics setGraphics(Graphics g) {
		if (g == null || (g = jsgraphics) == null) {
			g = jsgraphics = new JSGraphics2D(getCanvas());
			// set methods for HTMLCanvasContext2D that are just direct assignments
			// did not work /**
			// * @j2sNative
			// *
			// * g.ctx._setLineWidth = function(d) {this.lineWidth = d};
			// * g.ctx._setFont = function(f) {this.font = f};
			// * g.ctx._setFillStyle = function(s) {this.fillStyle = s};
			// * g.ctx._setStrokeStyle = function(s) {this.strokeStyle = s};
			// */
			// {}
			((JSGraphics2D) g).setWindowParameters(getWidth(), getHeight());
		}
		return g;
	}

	private void showAppletStatus(String status) {
		getAppletContext().showStatus(htmlName + " " + status);
	}

	private void showAppletException(Throwable t) {
		/**
		 * @j2sNative
		 * 
		 *            this.showAppletStatus("error " + (t.getMessage ?
		 *            t.getMessage() : t)); t.printStackTrace &&
		 *            t.printStackTrace();
		 */
		{
		}
		repaint();
	}

	/**
	 * 
	 * @param mode
	 * @return LOOP or DONE
	 */
	public int run1(int mode) {
		System.out.println("JSAP run1 mode " + mode + " " + nextStatus);
		boolean ok = false;
		switch (mode) {
		case JSThread.INIT:
			currentAppletSize.width = defaultAppletSize.width = getWidth();
			currentAppletSize.height = defaultAppletSize.height = getHeight();
			setLayout(new BorderLayout());
			nextStatus = APPLET_LOAD;
			ok = true;
			break;
		case JSThread.LOOP:
			switch (nextStatus) {
			case APPLET_LOAD:
				if (status != APPLET_UNINITIALIZED) {
					showAppletStatus("notdisposed");
					status = APPLET_ERROR;
					break;
				}
				System.out.println("JSAppletPanel runloader");
				runLoader(); // applet created here
				nextStatus = APPLET_INIT;
				ok = true;
				break;
			case APPLET_INIT:
				// AppletViewer "Restart" will jump from destroy method to
				// init, that is why we need to check status w/ APPLET_DESTROY
				if (status != APPLET_LOAD && status != APPLET_DESTROY) {
					showAppletStatus("notloaded");
					break;
				}
				System.out.println("JSAppletPanel init");
				applet.resize(defaultAppletSize);
				applet.init();
				// Need the default(fallback) font to be created in this AppContext
				Font f = getFont();
				if (f == null || "dialog".equals(f.getFamily().toLowerCase())// Locale.ENGLISH))
						&& f.getSize() == 12 && f.getStyle() == Font.PLAIN) {
					setFont(new Font(Font.DIALOG, Font.PLAIN, 12));
				}
				validate(); // SwingJS
				status = APPLET_INIT;
				showAppletStatus("initialized");
				nextStatus = APPLET_START;
				ok = true;
				break;
			case APPLET_START:
				if (status != APPLET_INIT && status != APPLET_STOP) {
					showAppletStatus("notstarted");
					status = APPLET_ERROR;
					break;
				}
				applet.getRootPane().addNotify();
				// force peer creation now
				System.out.println("JSAppletPanel start" + currentAppletSize);
				applet.resize(currentAppletSize);
				applet.start();
				status = APPLET_START;
				showAppletStatus("started");
				nextStatus = APPLET_READY;
				ok = true;
				break;
			case APPLET_READY:
				JSToolkit.readyCallback(appletName,fullName,applet,this);
				break;
			case APPLET_STOP:
				if (status == APPLET_START) {
					status = APPLET_STOP;
					applet.setVisible(false);
					applet.stop();
					showAppletStatus("stopped");
				} else {
					showAppletStatus("notstopped");
					status = APPLET_ERROR;
				}
				break;
			case APPLET_DESTROY:
				if (status == APPLET_STOP || status == APPLET_INIT) {
					status = APPLET_DESTROY;
					applet.destroy();
					showAppletStatus("destroyed");
				} else {
					showAppletStatus("notdestroyed");
					status = APPLET_ERROR;
				}
				break;
			case APPLET_DISPOSE:
				if (status == APPLET_DESTROY || status == APPLET_LOAD) {
					showAppletStatus("notdisposed");
					status = APPLET_ERROR;
				} else {
					status = APPLET_UNINITIALIZED;
					removeChild(applet);
					applet = null;
					showAppletStatus("disposed");
				}
				break;
			case APPLET_QUIT:
				break;
			default:
				System.out.println("unrecognized JSAppletPanel status: " + nextStatus);
				break;
			}
			break;
		default:
			System.out.println("unrecognized JSAppletThread mode: " + mode);
			break;
		}
		return (ok ? JSThread.LOOP : JSThread.DONE);
	}

	private void runLoader() {
		dispatchAppletEvent(APPLET_LOADING, null);
		status = APPLET_LOAD;
		String code = getCode();
		try {
			if (code == null) {
				System.err.println("runloader.err-- \"code\" must be specified.");
				throw new InstantiationException("\"code\" must be specified.");
			}
			applet = (JApplet) JSToolkit.getInstance(code);
			if (!(applet instanceof JApplet)) {
				System.out.println(code + "is not a JApplet!?");
				status = APPLET_ERROR;
			}
		} catch (InstantiationException e) {
			status = APPLET_ERROR;
			showAppletException(e);
			return;
		} catch (Exception e) {
			status = APPLET_ERROR;
			showAppletException(e);
			return;
		} catch (ThreadDeath e) {
			status = APPLET_ERROR;
			showAppletStatus("death");
			return;
		} catch (Error e) {
			status = APPLET_ERROR;
			showAppletException(e);
			return;
		} finally {
			// notify that loading is no longer going on
			dispatchAppletEvent(APPLET_LOADING_COMPLETED, null);
		}
		if (applet != null) {
			applet.setStub(this);
			applet.setVisible(false);
			add("Center", applet);
			showAppletStatus("loaded");
			validate();
		}
	}

}

File: site/swingjs/j2s/swingjs/JSAppletPanel.js
Clazz.declarePackage ("swingjs");
Clazz.load (["java.applet.AppletContext", "$.AppletStub", "java.awt.Panel", "swingjs.api.JSInterface", "java.awt.Dimension"], "swingjs.JSAppletPanel", ["java.lang.InstantiationException", "java.net.URL", "java.awt.BorderLayout", "$.Font", "$.Toolkit", "javax.swing.JApplet", "jssun.applet.AppletEvent", "$.AppletEventMulticaster", "swingjs.JSAppletThread", "$.JSGraphics2D", "$.JSMouse", "$.JSThreadGroup", "$.JSToolkit", "$.JSUtil", "swingjs.api.Interface"], function () {
c$ = Clazz.decorateAsClass (function () {
this.params = null;
this.html5Applet = null;
this.fullName = null;
this.appletCodeBase = null;
this.appletIdiomaBase = null;
this.appletDocumentBase = null;
this.maximumSize = 2147483647;
this.appletName = null;
this.syncId = null;
this.testAsync = false;
this.async = false;
this.strJavaVersion = null;
this.strJavaVendor = null;
this.display = null;
this.canvas = null;
this.jsgraphics = null;
this.applet = null;
this.defaultAppletSize = null;
this.currentAppletSize = null;
this.nextStatus = 0;
this.status = 0;
this.listeners = null;
this.mouse = null;
Clazz.instantialize (this, arguments);
}, swingjs, "JSAppletPanel", java.awt.Panel, [java.applet.AppletStub, java.applet.AppletContext, swingjs.api.JSInterface]);
Clazz.prepareFields (c$, function () {
this.defaultAppletSize =  new java.awt.Dimension (10, 10);
this.currentAppletSize =  new java.awt.Dimension (10, 10);
});
Clazz.makeConstructor (c$, 
function (params) {
Clazz.superConstructor (this, swingjs.JSAppletPanel, []);
this.set (params);
}, "java.util.Hashtable");
Clazz.defineMethod (c$, "set", 
 function (params) {
System.out.println ("JSAppletPanel initializing");
this.params = params;
this.htmlName = swingjs.JSUtil.split ("" + this.getParameter ("name"), "_object")[0];
this.appletName = swingjs.JSUtil.split (this.htmlName + "_", "_")[0];
this.syncId = this.getParameter ("syncId");
this.fullName = this.htmlName + "__" + this.syncId + "__";
params.put ("fullName", this.fullName);
var o = params.get ("codePath");
if (o == null) o = "../java/";
this.appletCodeBase = o.toString ();
this.appletIdiomaBase = this.appletCodeBase.substring (0, this.appletCodeBase.lastIndexOf ("/", this.appletCodeBase.length - 2) + 1) + "idioma";
o = params.get ("documentBase");
this.appletDocumentBase = (o == null ? "" : o.toString ());
if (params.containsKey ("maximumSize")) Math.max ((params.get ("maximumSize")).intValue (), 100);
this.async = (this.testAsync || params.containsKey ("async"));
var applet = null;
var javaver = "?";
{
if(self.Jmol) { applet =
Jmol._applets[this.htmlName.split("_object")[0]]; javaver =
Jmol._version; }
}this.html5Applet = applet;
this.strJavaVersion = javaver;
this.strJavaVendor = "Java2Script/Java 1.6 (HTML5)";
this.display = params.get ("display");
this.threadGroup =  new swingjs.JSThreadGroup (this.appletName);
this.myThread =  new swingjs.JSAppletThread (this, this.threadGroup, this.appletName);
(this.threadGroup).setHtmlApplet (this.html5Applet);
{
Jmol._applets[this.appletName + "_thread"] =
java.lang.Thread.thisThread = this.myThread;
}this.appContext = swingjs.JSToolkit.createNewAppContext ();
java.awt.Toolkit.getDefaultToolkit ();
try {
java.net.URL.setURLStreamHandlerFactory (swingjs.api.Interface.getInstance ("JU.AjaxURLStreamHandlerFactory", false));
} catch (e) {
}
System.out.println ("JSAppletPanel initialized");
}, "java.util.Hashtable");
Clazz.defineMethod (c$, "start", 
function () {
if (this.status == 0) this.myThread.start ();
 else this.showStatus ("already started");
});
Clazz.defineMethod (c$, "addAppletListener", 
function (l) {
this.listeners = jssun.applet.AppletEventMulticaster.add (this.listeners, l);
}, "jssun.applet.AppletListener");
Clazz.defineMethod (c$, "removeAppletListener", 
function (l) {
this.listeners = jssun.applet.AppletEventMulticaster.remove (this.listeners, l);
}, "jssun.applet.AppletListener");
Clazz.defineMethod (c$, "dispatchAppletEvent", 
function (id, argument) {
if (this.listeners != null) {
var evt =  new jssun.applet.AppletEvent (this, id, argument);
this.listeners.appletStateChanged (evt);
}}, "~N,~O");
Clazz.defineMethod (c$, "getCode", 
 function () {
return this.getParameter ("code");
});
Clazz.defineMethod (c$, "getCanvas", 
 function () {
return (this.canvas == null ? (this.canvas = this.html5Applet._getHtml5Canvas ()) : this.canvas);
});
Clazz.overrideMethod (c$, "isActive", 
function () {
return true;
});
Clazz.overrideMethod (c$, "appletResize", 
function (width, height) {
var currentSize =  new java.awt.Dimension (this.currentAppletSize.width, this.currentAppletSize.height);
this.currentAppletSize.width = width;
this.currentAppletSize.height = height;
this.applet.setBounds (0, 0, this.getWidth (), this.getHeight ());
this.applet.getRootPane ().setBounds (0, 0, this.getWidth (), this.getHeight ());
this.applet.getContentPane ().setBounds (0, 0, this.getWidth (), this.getHeight ());
(this.applet.getContentPane ()).revalidate ();
this.dispatchAppletEvent (51234, currentSize);
}, "~N,~N");
Clazz.overrideMethod (c$, "getDocumentBase", 
function () {
try {
return  new java.net.URL (this.params.get ("documentBase"));
} catch (e) {
if (Clazz.exceptionOf (e, java.net.MalformedURLException)) {
return null;
} else {
throw e;
}
}
});
Clazz.overrideMethod (c$, "getCodeBase", 
function () {
try {
return  new java.net.URL (this.params.get ("codePath"));
} catch (e) {
if (Clazz.exceptionOf (e, java.net.MalformedURLException)) {
return null;
} else {
throw e;
}
}
});
Clazz.overrideMethod (c$, "getParameter", 
function (name) {
var s = this.params.get (name);
System.out.println ("get parameter: " + name + " = " + s);
return (s == null ? null : "" + s);
}, "~S");
Clazz.overrideMethod (c$, "getAppletContext", 
function () {
return this;
});
Clazz.defineMethod (c$, "getHeight", 
function () {
return this.html5Applet._getHeight ();
});
Clazz.defineMethod (c$, "getWidth", 
function () {
return this.html5Applet._getWidth ();
});
Clazz.defineMethod (c$, "setBounds", 
function (x, y, width, height) {
this.reshape (x, y, width, height);
this.currentAppletSize.width = width;
this.currentAppletSize.height = height;
}, "~N,~N,~N,~N");
Clazz.overrideMethod (c$, "getImage", 
function (url) {
return null;
}, "java.net.URL");
Clazz.overrideMethod (c$, "getApplet", 
function (name) {
var applet = null;
{
applet = Jmol._applets[name]; applet && (applet =
applet._applet);
}return applet;
}, "~S");
Clazz.overrideMethod (c$, "getApplets", 
function () {
return null;
});
Clazz.defineMethod (c$, "showDocument", 
function (url) {
{
window.open(url.toString());
}}, "java.net.URL");
Clazz.defineMethod (c$, "showDocument", 
function (url, target) {
{
window.open(url.toString(), target);
}}, "java.net.URL,~S");
Clazz.defineMethod (c$, "showStatus", 
function (status) {
swingjs.JSToolkit.log (status);
{
Clazz._LoaderProgressMonitor.showStatus(status, true);
}}, "~S");
Clazz.overrideMethod (c$, "cacheFileByName", 
function (fileName, isAdd) {
return 0;
}, "~S,~B");
Clazz.overrideMethod (c$, "cachePut", 
function (key, data) {
}, "~S,~O");
Clazz.overrideMethod (c$, "destroy", 
function () {
});
Clazz.overrideMethod (c$, "getFullName", 
function () {
return this.fullName;
});
Clazz.overrideMethod (c$, "openFileAsyncSpecial", 
function (fileName, flags) {
}, "~S,~N");
Clazz.defineMethod (c$, "processMouseEvent", 
function (id, x, y, modifiers, time) {
this.getMouse ().processEvent (id, x, y, modifiers, time);
return false;
}, "~N,~N,~N,~N,~N");
Clazz.defineMethod (c$, "getMouse", 
 function () {
return (this.mouse == null ? this.mouse =  new swingjs.JSMouse (this) : this.mouse);
});
Clazz.overrideMethod (c$, "processTwoPointGesture", 
function (touches) {
this.getMouse ().processTwoPointGesture (touches);
}, "~A");
Clazz.overrideMethod (c$, "setDisplay", 
function (canvas) {
this.canvas = canvas;
}, "swingjs.api.HTML5Canvas");
Clazz.overrideMethod (c$, "setScreenDimension", 
function (width, height) {
this.setGraphics (this.jsgraphics = null);
if (this.applet != null) this.applet.resize (width, height);
}, "~N,~N");
Clazz.overrideMethod (c$, "setStatusDragDropped", 
function (mode, x, y, fileName) {
return false;
}, "~N,~N,~N,~S");
Clazz.overrideMethod (c$, "startHoverWatcher", 
function (enable) {
}, "~B");
Clazz.overrideMethod (c$, "paint", 
function (g) {
this.applet.paint (this.setGraphics (g));
}, "java.awt.Graphics");
Clazz.defineMethod (c$, "setGraphics", 
 function (g) {
return (g == null ? this.getGraphics () : g);
}, "java.awt.Graphics");
Clazz.overrideMethod (c$, "getGraphics", 
function () {
if (this.jsgraphics == null) {
this.jsgraphics =  new swingjs.JSGraphics2D (this.getCanvas ());
this.jsgraphics.setWindowParameters (this.getWidth (), this.getHeight ());
}return this.jsgraphics;
});
Clazz.defineMethod (c$, "showAppletStatus", 
 function (status) {
this.getAppletContext ().showStatus (this.htmlName + " " + status);
}, "~S");
Clazz.defineMethod (c$, "showAppletException", 
 function (t) {
{
this.showAppletStatus("error " + (t.getMessage ?
t.getMessage() : t)); t.printStackTrace &&
t.printStackTrace();
}this.repaint ();
}, "Throwable");
Clazz.defineMethod (c$, "run1", 
function (mode) {
var ok = false;
switch (mode) {
case 0:
this.currentAppletSize.width = this.defaultAppletSize.width = this.getWidth ();
this.currentAppletSize.height = this.defaultAppletSize.height = this.getHeight ();
this.setLayout ( new java.awt.BorderLayout ());
this.nextStatus = 1;
ok = true;
break;
case 1:
switch (this.nextStatus) {
case 1:
if (this.status != 0) {
this.showAppletStatus ("notdisposed");
this.status = 7;
break;
}System.out.println ("JSAppletPanel runloader");
this.runLoader ();
this.nextStatus = 2;
ok = true;
break;
case 2:
if (this.status != 1 && this.status != 5) {
this.showAppletStatus ("notloaded");
break;
}System.out.println ("JSAppletPanel init");
this.setFont ( new java.awt.Font ("Dialog", 0, 12));
this.applet.resize (this.defaultAppletSize);
this.applet.init ();
this.validate ();
this.status = 2;
this.showAppletStatus ("initialized");
this.nextStatus = 3;
ok = true;
break;
case 3:
if (this.status != 2 && this.status != 4) {
this.showAppletStatus ("notstarted");
this.status = 7;
break;
}this.applet.getRootPane ().addNotify ();
System.out.println ("JSAppletPanel start" + this.currentAppletSize);
this.applet.resize (this.currentAppletSize);
this.applet.start ();
this.status = 3;
this.showAppletStatus ("started");
this.nextStatus = 35;
ok = true;
break;
case 35:
swingjs.JSToolkit.readyCallback (this.appletName, this.fullName, this.applet, this);
break;
case 4:
if (this.status == 3) {
this.status = 4;
this.applet.setVisible (false);
this.applet.stop ();
this.showAppletStatus ("stopped");
} else {
this.showAppletStatus ("notstopped");
this.status = 7;
}break;
case 5:
if (this.status == 4 || this.status == 2) {
this.status = 5;
this.applet.destroy ();
this.showAppletStatus ("destroyed");
} else {
this.showAppletStatus ("notdestroyed");
this.status = 7;
}break;
case 75:
if (this.status == 5 || this.status == 1) {
this.showAppletStatus ("notdisposed");
this.status = 7;
} else {
this.status = 0;
this.removeChild (this.applet);
this.applet = null;
this.showAppletStatus ("disposed");
}break;
case 6:
break;
default:
System.out.println ("unrecognized JSAppletPanel status: " + this.nextStatus);
break;
}
break;
default:
System.out.println ("unrecognized JSAppletThread mode: " + mode);
break;
}
return (ok ? 1 : 2);
}, "~N");
Clazz.defineMethod (c$, "runLoader", 
 function () {
this.dispatchAppletEvent (51235, null);
this.status = 1;
var code = this.getCode ();
try {
if (code == null) {
System.err.println ("runloader.err-- \"code\" must be specified.");
throw  new InstantiationException ("\"code\" must be specified.");
}this.applet = swingjs.JSToolkit.getInstance (code);
if (this.applet == null) {
System.out.println (code + " could not be launched");
this.status = 7;
} else if (!(Clazz.instanceOf (this.applet, javax.swing.JApplet))) {
System.out.println (code + " is not a JApplet!?");
this.status = 7;
}} catch (e$$) {
if (Clazz.exceptionOf (e$$, InstantiationException)) {
var e = e$$;
{
this.status = 7;
this.showAppletException (e);
return;
}
} else if (Clazz.exceptionOf (e$$, Exception)) {
var e = e$$;
{
this.status = 7;
this.showAppletException (e);
return;
}
} else if (Clazz.exceptionOf (e$$, ThreadDeath)) {
var e = e$$;
{
this.status = 7;
this.showAppletStatus ("death");
return;
}
} else if (Clazz.exceptionOf (e$$, Error)) {
var e = e$$;
{
this.status = 7;
this.showAppletException (e);
return;
}
} else {
throw e$$;
}
} finally {
this.dispatchAppletEvent (51236, null);
}
if (this.applet != null) {
this.applet.setStub (this);
this.applet.setVisible (false);
this.add ("Center", this.applet);
this.applet.setDispatcher ();
this.applet.addNotify ();
this.applet.setVisible (true);
this.showAppletStatus ("loaded");
this.validate ();
}});
Clazz.defineMethod (c$, "getUI", 
function () {
return null;
});
Clazz.overrideMethod (c$, "repaintImpl", 
function (tm, x, y, width, height) {
if (this.applet.isVisible () && this.applet.getWidth () > 0 && this.applet.getHeight () > 0) {
this.applet.getContentPane ().repaint (tm, x, y, width, height);
}}, "~N,~N,~N,~N,~N");
Clazz.defineStatics (c$,
"APPLET_UNINITIALIZED", 0,
"APPLET_LOAD", 1,
"APPLET_INIT", 2,
"APPLET_START", 3,
"APPLET_READY", 35,
"APPLET_STOP", 4,
"APPLET_DESTROY", 5,
"APPLET_QUIT", 6,
"APPLET_ERROR", 7,
"APPLET_DISPOSE", 75,
"APPLET_RESIZE", 51234,
"APPLET_LOADING", 51235,
"APPLET_LOADING_COMPLETED", 51236);
});
